lib/vnet/install_service_windows.go

```diff
@@ -0,0 +1,316 @@
+	"golang.org/x/sys/windows/svc/mgr"
+
+	"github.com/gravitational/teleport/lib/windowsexec"
+)
+
+// InstallService installs the VNet windows service.
+//
+// Windows services are installed by the service manager, which takes a path to
+// the service executable. So that regular users are not able to overwrite the
+// executable at that path, we use a path under C:\Program Files\, which is not
+// writable by regular users by default.
+//
+// The username and logFile arguments are expected to be empty when this is
+// called by the regular user. The function will find the current username
+// and create a log file, then re-execute tsh with extra arguments setting
+// username and logFile.
+func InstallService(ctx context.Context, username, logFile string) (err error) {
+	// If not already running with elevated permissions, exec a child process of
+	// the current executable with the current args with `runas`.
+	if !windows.GetCurrentProcessToken().IsElevated() {
+		return trace.Wrap(elevateChildProcess(ctx),
+			"elevating process to install VNet Windows service")
+	}
+
+	if logFile == "" {
+		return trace.BadParameter("log-file is required")
+	}
+	defer func() {
+		// Write any errors to logFile so the parent process can read it.
+		if err != nil {
+			// Not really any point checking the error from WriteFile since
+			// noone will be able to read it.
+			os.WriteFile(logFile, []byte(err.Error()), 0644)
+		}
+	}()
+
+	if username == "" {
+		return trace.BadParameter("username is required")
+	}
+	u, err := user.Lookup(username)
+	if err != nil {
+		return trace.Wrap(err, "looking up user %s", username)
+	}
+
+	currentTshPath, err := os.Executable()
+	if err != nil {
+		return trace.Wrap(err, "getting current exe path")
+	}
+	currentWintunPath, err := wintunPath(currentTshPath)
+	if err != nil {
+		return trace.Wrap(err, "getting current wintun.dll path")
+	}
+
+	svcMgr, err := mgr.Connect()
+	if err != nil {
+		return trace.Wrap(err, "connecting to Windows service manager")
+	}
+
+	serviceName, serviceInstallDir := serviceNameAndInstallDir(username)
+	targetTshPath := filepath.Join(serviceInstallDir, "tsh.exe")
+	targetWintunPath := filepath.Join(serviceInstallDir, "wintun.dll")
+	if err := os.Mkdir(serviceInstallDir, 0600); err != nil && !errors.Is(err, fs.ErrExist) {
+		return trace.Wrap(err, "creating service installation directory %s", serviceInstallDir)
+	}
+	if err := copyFile(targetTshPath, currentTshPath); err != nil {
+		return trace.Wrap(err, "copying tsh.exe to service installation directory")
+	}
+	if err := copyFile(targetWintunPath, currentWintunPath); err != nil {
+		return trace.Wrap(err, "copying wintun.dll to service installation directory")
+	}
+
+	if existingSvc, err := svcMgr.OpenService(serviceName); err == nil {
+		// The service already exists, delete it and recreate in case
+		// configuration options have changed since it was originally installed.
+		if err := existingSvc.Delete(); err != nil {
```

Comment:
Does the service need to be deleted or just stopped and restarted?

Reply:
my idea here is to do a clean delete/recreate in case we change any paths or security parameters across versions

Reply:
I think using separate install directories per version solves this, and removes the need for any separate update and install logic

Reply:
I realized that the installer always runs the uninstaller before installing a new version, which will uninstall the service, so this is no longer necessary